bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcher.java

```diff
@@ -235,7 +255,15 @@ public void processResult(int rc, String path, Object ctx, List<String> children
      */
     public ArrayList<BookieSocketAddress> newEnsemble(int ensembleSize, int writeQuorumSize)
             throws BKNotEnoughBookiesException {
-        return placementPolicy.newEnsemble(ensembleSize, writeQuorumSize, EMPTY_SET);
+        try {
+            // we try to only get from the healthy bookies first
+            return placementPolicy.newEnsemble(ensembleSize, writeQuorumSize, quarantinedBookies.asMap().keySet());
```

Comment:
the key set will be modified in #newEnsembel(). It seems to me you have to create a new set from current quarantinedBookies's key set